### PR TITLE
MATH-1569 - Manual array copy.

### PR DESCRIPTION
**Pull request by arturobernalg**

### CODE REVIEW

**Comment by arturobernalg** [comment]
@aherbert WDYT?

File: commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java [hunk 1]
```diff
@@ -663,9 +663,7 @@ public Dfp getTwo() {
     /** Shift the mantissa left, and adjust the exponent to compensate.
      */
     protected void shiftLeft() {
-        for (int i = mant.length - 1; i > 0; i--) {
-            mant[i] = mant[i - 1];
-        }
+        if (mant.length - 1 >= 0) System.arraycopy(mant, 0, mant, 1, mant.length - 1);
```
Comment (aherbert) [inline commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java:666]: use `> 0` as arraycopy with a length of 0 is a wasted call to a native method. Add enclosing braces to fix checkstyle.

File: commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java [hunk 2]
```diff
@@ -675,9 +673,7 @@ uses shiftRight() */
     /** Shift the mantissa right, and adjust the exponent to compensate.
      */
     protected void shiftRight() {
-        for (int i = 0; i < mant.length - 1; i++) {
-            mant[i] = mant[i + 1];
-        }
+        if (mant.length - 1 >= 0) System.arraycopy(mant, 1, mant, 0, mant.length - 1);
```
Comment (aherbert) [inline commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java:676]: `> 0`

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/function/Logit.java [hunk 3]
```diff
@@ -182,9 +182,7 @@ public DerivativeStructure value(final DerivativeStructure t)
             //  for x close to hi the signs will always be +inf)
             // this is probably overkill, since the call to compose at the end
             // of the method will transform most infinities into NaN ...
-            for (int i = 2; i < f.length; ++i) {
-                f[i] = f[i - 2];
-            }
+            if (f.length - 2 >= 0) System.arraycopy(f, 0, f, 2, f.length - 2);
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/function/Logit.java:185]: `> 0`

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/ode/nonstiff/AdamsNordsieckFieldTransformer.java [hunk 4]
```diff
@@ -170,10 +170,8 @@ private AdamsNordsieckFieldTransformer(final Field<T> field, final int n) {
         // Nordsieck to multistep, then shifting rows to represent step advance
         // then applying inverse transform
         T[][] shiftedP = bigP.getData();
-        for (int i = shiftedP.length - 1; i > 0; --i) {
-            // shift rows
-            shiftedP[i] = shiftedP[i - 1];
-        }
+        // shift rows
+        if (shiftedP.length - 1 >= 0) System.arraycopy(shiftedP, 0, shiftedP, 1, shiftedP.length - 1);
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/ode/nonstiff/AdamsNordsieckFieldTransformer.java:174]: `> 0`

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/optim/nonlinear/scalar/noderiv/CMAESOptimizer.java [hunk 5]
```diff
@@ -752,9 +752,7 @@ private void updateBD(double negccov) {
      * @param val Current best fitness value.
      */
     private static void push(double[] vals, double val) {
-        for (int i = vals.length-1; i > 0; i--) {
-            vals[i] = vals[i-1];
-        }
+        if (vals.length - 1 >= 0) System.arraycopy(vals, 0, vals, 1, vals.length - 1);
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/optim/nonlinear/scalar/noderiv/CMAESOptimizer.java:755]: `> 0`

**Review by aherbert — CHANGES_REQUESTED** [review]
Checkstyle is failing this too. The changes to use arraycopy look good but please add the missing enclosing braces.

**Comment by arturobernalg** [comment]
@aherbert  all the remark changed.

File: commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java [hunk 6]
```diff
@@ -663,8 +663,8 @@ public Dfp getTwo() {
     /** Shift the mantissa left, and adjust the exponent to compensate.
      */
     protected void shiftLeft() {
-        for (int i = mant.length - 1; i > 0; i--) {
-            mant[i] = mant[i - 1];
+        if (mant.length - 1 >= 0) {
```
Comment (aherbert) [inline commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java:666]: `> 0`

**Review by aherbert — CHANGES_REQUESTED** [review]

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/differentiation/DSCompiler.java [hunk 7]
```diff
@@ -1461,9 +1461,7 @@ public void cosh(final double[] operand, final int operandOffset,
         function[0] = JdkMath.cosh(operand[operandOffset]);
         if (order > 0) {
             function[1] = JdkMath.sinh(operand[operandOffset]);
-            for (int i = 2; i <= order; ++i) {
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/differentiation/DSCompiler.java:1464]: This is wrong

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/differentiation/DSCompiler.java [hunk 8]
```diff
@@ -1486,9 +1484,7 @@ public void sinh(final double[] operand, final int operandOffset,
         function[0] = JdkMath.sinh(operand[operandOffset]);
         if (order > 0) {
             function[1] = JdkMath.cosh(operand[operandOffset]);
-            for (int i = 2; i <= order; ++i) {
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/differentiation/DSCompiler.java:1489]: This is wrong

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/function/Logit.java [hunk 9]
```diff
@@ -182,8 +182,8 @@ public DerivativeStructure value(final DerivativeStructure t)
             //  for x close to hi the signs will always be +inf)
             // this is probably overkill, since the call to compose at the end
             // of the method will transform most infinities into NaN ...
-            for (int i = 2; i < f.length; ++i) {
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/function/Logit.java:185]: This is wrong. f[0 and f[1] have been set and should be duplicated through the array.

**Review by aherbert — CHANGES_REQUESTED** [review]
I noticed the CI build failure on the same tests in the differentiation package. This is because some of the loop copy statements are duplicating the first 2 elements through the rest of the array. This is different from a system arraycopy which will copy the first two elements to the next two elements. But the downstream elements are not duplicates of the first two, they are copies of the two previous elements in the original array:

```
data = [a, b, c, d, e, f]

for (int i = 2; i < data.length; ++i) {
    data[i] = data[i - 2];
}
// data = [a, b, a, b, a, b]

arraycopy(data, 0, data, 2, data.length - 2);
// data = [a, b, a, b, c, d]
```

**Comment by arturobernalg** [comment]
> I noticed the CI build failure on the same tests in the differentiation package. This is because some of the loop copy statements are duplicating the first 2 elements through the rest of the array. This is different from a system arraycopy which will copy the first two elements to the next two elements. But the downstream elements are not duplicates of the first two, they are copies of the two previous elements in the original array:
> 
> ```
> data = [a, b, c, d, e, f]
> 
> for (int i = 2; i < data.length; ++i) {
>     data[i] = data[i - 2];
> }
> // data = [a, b, a, b, a, b]
> 
> arraycopy(data, 0, data, 2, data.length - 2);
> // data = [a, b, a, b, c, d]
> ```

Now I'm confused. What soul I do with this PR?

**Comment by aherbert** [comment]
Where I have put a note that 'This is wrong' you should revert the change. The arraycopy does not duplicate the first two elements through the array in an alternating pattern; the original loop does. This is why the tests fail. It should be reverted.

**Comment by arturobernalg** [comment]
> Where I have put a note that 'This is wrong' you should revert the change. The arraycopy does not duplicate the first two elements through the array in an alternating pattern; the original loop does. This is why the tests fail. It should be reverted.

OK. done.

**Comment by aherbert** [comment]
This is failing CI on checkstyle. You have removed 2 lines from the
`org.apache.commons.math4.legacy.fitting.leastsquares.LevenbergMarquardtOptimizer`.

There is this suppression for the legacy module:
```
<suppress checks="UnnecessaryParentheses" files=".*[/\\]LevenbergMarquardtOptimizer\.java$" lines="527,741" />
```

But the CI fails with unnecessary parentheses at line 525 and 739. So you have to update the lines in:
`src/main/resources/checkstyle/checkstyle-suppressions-legacy.xml`

**Comment by aherbert** [comment]
I did not say to remove the unnecessary parentheses in the `LevenbergMarquardtOptimizer`. I said to update the checkstyle suppression file to correctly suppress this checkstyle violation. We want to keep the unnecessary parentheses.

**Comment by arturobernalg** [comment]
> I did not say to remove the unnecessary parentheses in the `LevenbergMarquardtOptimizer`. I said to update the checkstyle suppression file to correctly suppress this checkstyle violation. We want to keep the unnecessary parentheses.

Ok, change it
TY

File: src/main/resources/checkstyle/checkstyle-suppressions-legacy.xml [hunk 10]
```diff
@@ -44,7 +44,7 @@
   <suppress checks="UnnecessaryParentheses" files=".*[/\\]SparseGradient\.java$" lines="385,396" />
   <suppress checks="UnnecessaryParentheses" files=".*[/\\]DerivativeStructure\.java$" lines="468,481" />
   <suppress checks="UnnecessaryParentheses" files=".*[/\\]SimpleCurveFitter\.java$" lines="321-322" />
-  <suppress checks="UnnecessaryParentheses" files=".*[/\\]LevenbergMarquardtOptimizer\.java$" lines="527,741" />
+  <suppress checks="UnnecessaryParentheses" files=".*[/\\]LevenbergMarquardtOptimizer\.java$" lines="525,527,741" />
```
Comment (aherbert) [inline src/main/resources/checkstyle/checkstyle-suppressions-legacy.xml:47]: Since you removed two lines above `527` this should be `"525,739"`. Please change and verify with `mvn checkstyle:check` in this module.

Comment (arturobernalg) [reply src/main/resources/checkstyle/checkstyle-suppressions-legacy.xml]: done.

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/fitting/leastsquares/LevenbergMarquardtOptimizer.java [hunk 11]
```diff
@@ -738,9 +736,9 @@ private double determineLMParameter(double[] qy, double delta, double[] diag,
             // if the function is small enough, accept the current value
             // of lmPar, also test for the exceptional cases where parl is zero
             if (JdkMath.abs(fp) <= 0.1 * delta ||
-                (parl == 0 &&
+                 parl == 0 &&
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/fitting/leastsquares/LevenbergMarquardtOptimizer.java:739]: Now you have fixed the checkstyle suppressions file, you can revert the changes to the parentheses

File: commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/solvers/LaguerreSolver.java [hunk 12]
```diff
@@ -274,9 +274,7 @@ public Complex[] solveAll(Complex[] coefficients, Complex initial)
             }
             // Coefficients for deflated polynomial.
             final Complex[] c = new Complex[n + 1];
-            for (int i = 0; i <= n; i++) {
-                c[i] = coefficients[i];
-            }
+            System.arraycopy(coefficients, 0, c, 0, n + 1);
```
Comment (aherbert) [inline commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/solvers/LaguerreSolver.java:277]: I think this may be better as:
```Java
final Complex[] c = coefficients.clone();
```

File: commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java (hunk 6 above)
Comment (aherbert) [reply commons-math-legacy-core/src/main/java/org/apache/commons/math4/legacy/core/dfp/Dfp.java:666]: This is still not updated to use ` > 0`

**Review by aherbert — CHANGES_REQUESTED** [review]
Nearly there. 1 small change not yet done, 1 reversion, and a new change to use array.clone().

Thanks.
